README
Graphics-Framebuffer

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

FURTHER TEST SCRIPTS

To test the installation properly.  Log into the text console (not X).
Go to the 't' directory and run 'frtest.pl'.  This basically calls
most of the features of the module.

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Graphics::Framebuffer

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Graphics-Framebuffer

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Graphics-Framebuffer

    CPAN Ratings
        http://cpanratings.perl.org/d/Graphics-Framebuffer

    Search CPAN
        http://search.cpan.org/dist/Graphics-Framebuffer/


LICENSE AND COPYRIGHT

Copyright (C) 2013 Richard Kelsch

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Graphics-Framebuffer

1.00    2004
        First version, released privately only, as part of a Perl
		media player project called "SuperSparky".

4.02    Dec 29, 2013
        Public CPAN release.  This is considerably more advanced
		than the one distributed as part of the media player.

4.03    March 20, 2014
        Minor changes to the CPAN package

4.05    April 14, 2014
        Documentation updates.  POD section revamped.

4.06    July 19, 2014
        Added better testing
        
        Added framebuffer emulation mode for test running on systems
        without a hardware framebuffer.  It draws to a 640x480x32
        virtual screen in memory.  You can dump this screen to a file.
        
4.07    July 23, 2014
        Fixed calculation error in string size allocation for emulation
        mode.
        
        Fixed fill 'flood' typo.
        
        Fixed pixel class documentation.  It was incorrect.

        Added more tests for most used classes.

4.08    July 25, 2014
        Added additional information to the documentation for the new
        emulation mode.  No actual code changes made.

4.09    November 26, 2014
        Added more FBIOCTL flags to make ready for future hardware
        accellerated updates.  No promises here, but heck, this
        entire module was an exercise in "can it be done in Perl?"
        
        Also fixed a code comment that was mislabeled.



t/frtest.pl
#!/usr/bin/perl

$| = 1;

use strict;
use Switch;

use threads;
# use Data::Dumper::Simple;
use Graphics::Framebuffer;

my @framebuffer;
my $Threads = 4; # Yep, threads can be used... carefully.
foreach my $thr (1..$Threads) {
    push(@framebuffer,Graphics::Framebuffer->new());
}
my $screen_width = $framebuffer[0]->{'XRES'};
my $screen_height = $framebuffer[0]->{'YRES'};
# print STDERR Dumper($framebuffer),"\n";
$framebuffer[0]->cls();
my $cls = int(rand($screen_width));
foreach my $page (1..$Threads) {
    threads->create(
        sub {
            my $Page = shift;
            while(1) {
                attract($Page);
            }
        },
        $page - 1
    );
}
while(1) {
    threads->yield();
}



##############################################################################
##                             ATTRACT MODE                                 ##
##############################################################################
# Remeniscent of the "Atract Mode" of the old Atari 8 bit computers, this    #
# mode merely puts random patterns on the screen.                            #
##############################################################################

sub attract {
    my $Page = shift;
    my $red = int(rand(256));
    my $grn = int(rand(256));
    my $blu = int(rand(256));
    my $x   = int(rand($screen_width));
    my $y   = int(rand($screen_height));
    my $w   = int(rand($screen_width/4));
    my $h   = int(rand($screen_height/4));
    my $rx  = int(rand($screen_width/4));
    my $ry  = int(rand($screen_height/4));
    my $sd  = int(rand(360));
    my $ed  = int(rand(360));
    my $gr  = (rand(6)/10) + .5;
    my $mode = int(rand(5));
    my $type = int(rand(7));
    my $size = int(rand(3));
    $framebuffer[$Page]->cls() if ($x >= ($cls - 4) && $x <= ($cls + 4));

    $framebuffer[$Page]->set_color({'red' => $red,'green' => $grn,'blue' => $blu});
    $framebuffer[$Page]->draw_mode($mode);
    $framebuffer[$Page]->clip_rset({'x' => 0,'y' => 0,'width' => $screen_width-1,'height' => $screen_height-1});
    switch ($type) {
        case 0 {
            $framebuffer[$Page]->plot({'x' => $x,'y' => $y,'pixel_size' => 1});
        }
        case 1 {
            $framebuffer[$Page]->plot({'x' => $x,'y' => $y,'pixel_size' => $size});
            $framebuffer[$Page]->drawto({'x' => $w,'y' => $h,'pixel_size' => $size});
        }
        case 2 {
            $framebuffer[$Page]->circle({'x' => $x,'y' => $y,'radius' => $rx,'filled' => int(rand(2)),'pixel_size' => $size});
        }
        case 3 {
            $framebuffer[$Page]->ellipse({'x' => $x,'y' => $y,'xradius' => $rx,'yradius' => $ry,'filled' => int(rand(2)),'factor' => 1,'pixel_size' => $size});
        }
        case 4 {
            $framebuffer[$Page]->rbox({'x' => $x,'y' => $y,'width' => $w,'height' => $h,'filled' => (int(rand(2))),'pixel_size' => $size});
        }
        case 5 {
            $framebuffer[$Page]->draw_arc({'x' => $x,'y' => $y,'radius' => $ry,'start_degrees' => $sd,'end_degrees' => $ed,'granularity' => $gr,'mode' => (int(rand(3))),'pixel_size' => $size});
        }
        case 6 {
            my @poly;
            foreach my $count (0..int(rand(10))) {
                push(@poly,int(rand($screen_width)));
                push(@poly,int(rand($screen_height)));
            }
            $framebuffer[$Page]->polygon({'pixel_size' => $size,'coordinates' => \@poly});
        }
        else { # This can be a large memory hog.  Disabled for the moment.
            $framebuffer[$Page]->clip_set({'x' => $x-50,'y' => $y-50,'xx' => $x+50,'yy' => $y+50});
            fill({'x' => $x,'y' => $y});
        }
    }
    $framebuffer[$Page]->clip_reset();
    return($Page);
}

##############################################################################
##                DEBUGGING AND MANUALLY EXECUTED ROUTINES                  ##
##############################################################################
# These routines are mainly used for debugging and ordinary coding purposes. #
# Some are not called unless coded into the system, and therefore are        #
# orphans during normal operation.                                           #
##############################################################################

sub chars {
    my $counter = 0;
    blank_screen();
    locate(1,1);
    for (my $count=129;$count<255;$count++) {
        print chr(9) . "$count:" . chr($count);
        if ($counter == 4) {
            print "\n";
            $counter = 0;
        } else {
            $counter++;
        }
    }
}

__END__

__C__



examples/frtest.pl
#!/usr/bin/perl

# This tests the various methods of the Graphics::Framebuffer module
# and shows the only way threads will work with it.

use strict;
use Switch;

use threads;
use Graphics::Framebuffer;

my @framebuffer;
my $Threads = 4;
foreach my $thr (1..$Threads) {
    push(@framebuffer,Graphics::Framebuffer->new());
}
my $screen_width = $framebuffer[0]->{'XRES'};
my $screen_height = $framebuffer[0]->{'YRES'};

$framebuffer[0]->cls();
my $cls = int(rand($screen_width));
foreach my $page (1..$Threads) {
    threads->create(
        sub {
            my $Page = shift;
            if (defined($framebuffer[$Page]->{'ERROR'})) {
                print STDERR "Thread $Page, $framebuffer[$Page]->{'ERROR'}\n";
            }
            while(1) {
                attract($Page);
            }
        },
        $page - 1
    );
}
while(1) {
    threads->yield();
}


##############################################################################
##                             ATTRACT MODE                                 ##
##############################################################################
# Remeniscent of the "Atract Mode" of the old Atari 8 bit computers, this    #
# mode merely puts random patterns on the screen.                            #
##############################################################################

sub attract {
    my $Page = shift;
    my $red = int(rand(256));
    my $grn = int(rand(256));
    my $blu = int(rand(256));
    my $x   = int(rand($screen_width));
    my $y   = int(rand($screen_height));
    my $w   = int(rand($screen_width/4));
    my $h   = int(rand($screen_height/4));
    my $rx  = int(rand($screen_width/4));
    my $ry  = int(rand($screen_height/4));
    my $sd  = int(rand(360));
    my $ed  = int(rand(360));
    my $gr  = (rand(6)/10) + .5;
    my $mode = int(rand(5));
    my $type = int(rand(7));
    my $size = int(rand(3));
    $framebuffer[$Page]->cls() if ($x >= ($cls - 4) && $x <= ($cls + 4));

    $framebuffer[$Page]->set_color({'red' => $red,'green' => $grn,'blue' => $blu});
    $framebuffer[$Page]->draw_mode($mode);
    $framebuffer[$Page]->clip_rset({'x' => 0,'y' => 0,'width' => $screen_width-1,'height' => $screen_height-1});
    switch ($type) {
        case 0 {
            $framebuffer[$Page]->plot({'x' => $x,'y' => $y,'pixel_size' => 1});
        }
        case 1 {
            $framebuffer[$Page]->plot({'x' => $x,'y' => $y,'pixel_size' => $size});
            $framebuffer[$Page]->drawto({'x' => $w,'y' => $h,'pixel_size' => $size});
        }
        case 2 {
            $framebuffer[$Page]->circle({'x' => $x,'y' => $y,'radius' => $rx,'filled' => int(rand(2)),'pixel_size' => $size});
        }
        case 3 {
            $framebuffer[$Page]->ellipse({'x' => $x,'y' => $y,'xradius' => $rx,'yradius' => $ry,'filled' => int(rand(2)),'factor' => 1,'pixel_size' => $size});
        }
        case 4 {
            $framebuffer[$Page]->rbox({'x' => $x,'y' => $y,'width' => $w,'height' => $h,'filled' => (int(rand(2))),'pixel_size' => $size});
        }
        case 5 {
            $framebuffer[$Page]->draw_arc({'x' => $x,'y' => $y,'radius' => $ry,'start_degrees' => $sd,'end_degrees' => $ed,'granularity' => $gr,'mode' => (int(rand(3))),'pixel_size' => $size});
        }
        case 6 {
            my @poly;
            foreach my $count (0..int(rand(10))) {
                push(@poly,int(rand($screen_width)));
                push(@poly,int(rand($screen_height)));
            }
            $framebuffer[$Page]->polygon({'pixel_size' => $size,'coordinates' => \@poly});
        }
        else { # This can be a large memory hog.  Disabled for the moment.
            $framebuffer[$Page]->clip_set({'x' => $x-50,'y' => $y-50,'xx' => $x+50,'yy' => $y+50});
            fill({'x' => $x,'y' => $y});
        }
    }
    $framebuffer[$Page]->clip_reset();
    return($Page);
}

__END__

__C__

